crowbar_framework/app/controllers/api/restart_management_controller.rb

```diff
@@ -0,0 +1,159 @@
+  api_version "2.0"
+  def restarts
+    if request.post?
+      restarts_post
+    else
+      restarts_get
+    end
+  end
+
+  private
+
+  # Gets a list of the barclamps and extract the values from their nodes
+  # this ignores the roles and groups together all the nodes under a barclamp
+  # this is because we use a barclamp-level service restart checks instead of
+  # doing it service level, which is more cumbersome and requires us to know a list of services
+  # beforehand
+  def restarts_get
+    # maybe we want to blacklist some barclamps from the restart list?
+    blacklist = []
+    barclamps = BarclampCatalog.members("openstack").keys.reject { |m| blacklist.include? m }
+    # Chef query search to find all nodes that have requires_restart attribute not empty
+    nodes_with_restarts = NodeObject.find("requires_restart:*")
+    @final_list = {}
+    nodes_with_restarts.each do |node|
+      @final_list[node[:fqdn]] = { alias: node.alias }
+      barclamps.each do |barclamp|
+        # skip if there is not attributes for the barclamp or the requires_restart is empty
+        # why node.node.fetch? Dont ask me, ask the chef creators why this works
+        # https://github.com/SUSE-Cloud/chef/blob/10-stable-suse/chef/lib/chef/node.rb#L187
+        requires_restart = node.node.fetch(barclamp.to_s, {}).fetch("requires_restart", {})
+        next if requires_restart.empty?
+        @final_list[node[:fqdn]].update(barclamp => requires_restart)
+      end
+    end
+
+    render json: @final_list
+  end
+
+  def restarts_post
+    params.require(:node)
+    params.require(:service)
+
+    node_name = params[:node]
+    service = params[:service]
+    barclamps = BarclampCatalog.members("openstack").keys
+
+    # Search for the exact node with the provided params. This should yield an unique node
+    node = NodeObject.find("name:#{node_name} AND requires_restart:#{service}").first
+
+    # validate that node exists
+    raise Crowbar::Error::NotFound if node.nil?
+
+    barclamps.each do |barclamp|
+      next unless node.key? barclamp
+      next unless node[barclamp].key? :requires_restart
+      next unless node[barclamp][:requires_restart].key? service
+      node[barclamp][:requires_restart].delete(service)
+    end
+
+    node.save
+
+    head :ok
+  end
+
+  def configuration_get
+    # If its a GET call, just show a list of the barclamps disallow_flag status
+    @final_list = Crowbar::DataBagConfig.get_or_create_databag_item(
+      "crowbar-config", "disallow_restart"
+    )
+    @final_list.raw_data.delete("id")
+    render json: @final_list.raw_data
+  end
+
+  def configuration_post
+    params.require(:disallow_restart)
+    params.require(:barclamp)
+
+    disallow_restart = params[:disallow_restart] == "true" ? true : false
+    barclamp = params[:barclamp]
+
+    # validate that the barclamp exists
+    raise Crowbar::Error::NotFound unless BarclampCatalog.members("openstack").include? barclamp
```

Comment:
Any reason to restrict that to openstack barclamps?

Reply:
The card criteria was focused on openstack services (actually just on neutron first, then nova) and it look easier to me to first focus on the openstack services. I mean, do we want to avoid restarts of the dns or network services? Not counting that, for example the network barclamp does a lot of system managing outside of chef resources.

Reply:
I do think there's no reason to restrict this to openstack barclamps, but it's probably okay-ish. No reason to block on this.